### PR TITLE
Migrate to Swift 4's Decodable APIs

### PR DESCRIPTION
**Pull request by Palleas**
- [x] Add CI target for Swift 4 in travis (cf #88)
- [x] Use Swift 4 Codable protocol
- [x] Update tests
- [x] Remove duplicate `Result` submodule

Let me know if I forgot something in the list of stuff to do!

### CODE REVIEW

**Comment by Palleas** [comment]
screen shot 2017-06-17 at 5 29 28 pm

Getting there!

**Comment by Palleas** [comment]
I'm running the unit tests now and I have an issue. Basically since we treat `id` in all the resources as `String` while the JSON returns `Int`, the decoding fails. The only solution I've found so far is doing is for `Release`:

```swift
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let id = try container.decode(Int.self, forKey: .id)
        self.id = "\(id)"
        self.tag = try container.decode(String.self, forKey: .tag)
        self.url = try container.decode(URL.self, forKey: .url)
        self.name = try container.decode(String.self, forKey: .tag)
        self.isDraft = try container.decode(Bool.self, forKey: .isDraft)
        self.isPrerelease = try container.decode(Bool.self, forKey: .isPrerelease)
        self.assets = try container.decode([Asset].self, forKey: .assets)
    }
```

Do you see a different approach we could take with this? Right now the only solution I see is going back to treating `id` as `Int`.  ¯\\_(ツ)_/¯

**Comment by mdiep** [comment]
I think that's likely the only solution. I don't think the built-in JSON encoding is as flexible.

**Comment by mdiep** [comment]
On second thought: I think we should add a property type.

```swift
extension Release {
    public struct ID {
        private let id: Int
        public var string: String {
            return "\(id)"
        }
    }
}
```

That might make decoding easier?

**Comment by Palleas** [comment]
@mdiep Ha! That's exactly what we were talking about in Slack, Cf: https://gist.github.com/Palleas/645b6c30306603feaab148c9e167a307

Although since it will happen for all the Resources, maybe we need to move it to a higher level?

**Comment by mdiep** [comment]
I think it makes sense to have a per-type `ID` because it adds additional type safety. You can't accidentally pass an ID for the wrong type of resource to an API.

But if it made sense, we could add a top-level protocol that they all conform to.

File: Sources/Tentacle/Client.swift [hunk 1]
```diff
@@ -26,8 +27,8 @@ extension URL {
     
     internal init<Value>(_ server: Server, _ request: Request<Value>, page: UInt? = nil, perPage: UInt? = nil) {
         let queryItems = [ ("page", page), ("per_page", perPage) ]
-            .filter { _, value in value != nil }
-            .map { name, value in URLQueryItem(name: name, value: "\(value!)") }
+            .filter { arg -> Bool in let (_, value) = arg; return value != nil }
+            .map { arg -> URLQueryItem in let (name, value) = arg; return URLQueryItem(name: name, value: "\(value!)") }
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: If we move to a newer Swift build, we should be able to revert this change since SE-0110 was reverted.

File: Sources/Tentacle/Client.swift [hunk 2]
```diff
@@ -188,51 +189,45 @@ public final class Client {
     }
 
     /// Fetch a request from the API.
-    private func execute<Value>(_ request: Request<Value>, page: UInt?, perPage: UInt?) -> SignalProducer<(Response, Any), Error> {
-        return urlSession
+    private func execute<Value: Decodable>(_ request: Request<Value>, page: UInt?, perPage: UInt?) -> SignalProducer<(Response, Data), Error> {
+        let s = urlSession
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: Should be able to remove this local variable and go back to `return`ing directly?

File: Sources/Tentacle/Client.swift [hunk 3]
```diff
@@ -188,51 +189,45 @@ public final class Client {
...
+            let httpResponse = response as! HTTPURLResponse
+            let githubResponse = Response(headerFields: httpResponse.allHeaderFields as! [String: String])
+
+            if httpResponse.statusCode == 404 {
+                return .failure(.doesNotExist)
+            }
+
+            if httpResponse.statusCode >= 400 && httpResponse.statusCode < 600 {
+                let error: Result<GitHubError, Error> = decode(data).mapError { Error.jsonDecodingError($0) }
+                return .failure(.apiError(httpResponse.statusCode, githubResponse, error.value!))
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: This `!` isn't really safe. We should go back to the previous structure using `flatMap`.

File: Sources/Tentacle/Client.swift [hunk 4]
```diff
@@ -188,51 +189,45 @@ public final class Client {
...
-                    .map { resource in
-                        (response, resource)
-                    }
-                    .mapError(Error.jsonDecodingError)
-            }
+    ) -> SignalProducer<(Response, Resource), Error> {
+        let s: SignalProducer<(Response, Data), Error> = execute(request, page: nil, perPage: nil)
+
+        return s.attemptMap({ (arg) -> Result<(Response, Resource), Client.Error> in
+            let (response, data) = arg
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: Same here WRT SE-0110.

File: Sources/Tentacle/Client.swift [hunk 5]
```diff
@@ -188,51 +189,45 @@ public final class Client {
...
-        return execute(request, page: nil, perPage: nil)
-            .attemptMap { response, JSON in
-                return decode(JSON)
-                    .map { resource in
-                        (response, resource)
-                    }
-                    .mapError(Error.jsonDecodingError)
-            }
+    ) -> SignalProducer<(Response, Resource), Error> {
+        let s: SignalProducer<(Response, Data), Error> = execute(request, page: nil, perPage: nil)
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: Should be able to remove `s` and just `return`.

File: Sources/Tentacle/Client.swift [hunk 6]
```diff
@@ -241,20 +236,28 @@ public final class Client {
...
+
+        return s
+            .attemptMap({ (arg) -> Result<(Response, [Resource]), Client.Error> in
+                let (response, data) = arg
+                return decodeList(data)
+                    .map { (response, $0) }
+                    .mapError { Error.jsonDecodingError($0) }
+            })
+            .flatMap(.concat, { (arg) -> SignalProducer<(Response, [Resource]), Error> in
+                let (response, _) = arg
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: Lots of SE-0110. 😞

File: Sources/Tentacle/Client.swift [hunk 7]
```diff
@@ -241,20 +236,28 @@ public final class Client {
...
+
+                let current = SignalProducer<(Response, [Resource]), Error>(value: arg)
+                guard let _ = response.links["next"] else {
+                    return current.concat(.empty)
+                }
+
+                let s = self.execute(request, page: nextPage, perPage: perPage)
+                return current.concat(s)
+            })
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: We should use trailing closure syntax where possible.

File: Sources/Tentacle/Client.swift [hunk 8]
```diff
@@ -268,7 +271,7 @@ extension Client.Error: Hashable {
             return (error1 as NSError) == (error2 as NSError)
 
         case let (.jsonDecodingError(error1), .jsonDecodingError(error2)):
-            return error1 == error2
+            return error1.errorCode == error2.errorCode // FIXME
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: 😕 

What properties are available on `error1` and `error2`?

File: Sources/Tentacle/Comment.swift [hunk 9]
```diff
@@ -20,10 +17,10 @@ extension Repository {
     }
 }
 
-public struct Comment: CustomStringConvertible {
+public struct Comment: CustomStringConvertible, ResourceType, Identifiable {
 
     /// The id of the issue
-    public let id: String
+    public let id: ID<Comment>
```
Comment (mdiep) [inline Sources/Tentacle/Comment.swift]: 👍

File: Sources/Tentacle/Homepage.swift [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+//  Homepage.swift
+//  Tentacle
+//
+//  Created by Romain Pouclet on 2017-06-20.
+//  Copyright © 2017 Matt Diephouse. All rights reserved.
+//
+
+import Foundation
+
+public struct Homepage: Decodable {
```
Comment (mdiep) [inline Sources/Tentacle/Homepage.swift]: Maybe make this an enum instead?

```swift
enum Homepage {
    case url(URL)
    case string(String)
}
```

File: Sources/Tentacle/Label.swift [hunk 11]
```diff
@@ -7,17 +7,30 @@
...
+
+    public init(name: String, color: Color) {
+        self.name = name
+        self.color = color
+    }
+
+    private enum CodingKeys: String, CodingKey {
+        case name
+        case color
+    }
```
Comment (mdiep) [inline Sources/Tentacle/Label.swift:33]: Do we need these since they match what would be generated?

File: Tests/TentacleTests/FileResponseTests.swift [hunk 12]
```diff
@@ -7,57 +7,56 @@
...
+//        )
+//
+//        let expected = FileResponse(content: content, commit: commit)
+//
+//        guard case let .success(decoded) = FileResponse.decode(json) else {
+//            XCTFail()
+//            return
+//        }
+//
+//        XCTAssertEqual(expected, decoded)
```
Comment (mdiep) [inline Tests/TentacleTests/FileResponseTests.swift]: 🤔

File: Tests/TentacleTests/FileTests.swift [hunk 13]
```diff
@@ -8,55 +8,54 @@
...
+//            branch: nil
+//        )
+//
+//        let expected: JSON = .object([
+//            "message": .string("Added file"),
+//            "content": .string("This is the content of my file".data(using: .utf8)!.base64EncodedString()),
+//        ])
+//
+//        let encoded = file.encode()
+//        XCTAssertEqual(expected, encoded)
```
Comment (mdiep) [inline Tests/TentacleTests/FileTests.swift]: 🤔

File: Tests/TentacleTests/FileTests.swift [hunk 14]
```diff
@@ -8,55 +8,54 @@
...
+//            ]),
+//            "author": .object([
+//                "name": .string("Romain Pouclet"),
+//                "email": .string("romain.pouclet@gmail.com")
+//            ]),
+//            "branch": .string("master")
+//        ])
+//
+//        let encoded = file.encode()
+//        XCTAssertEqual(expected, encoded)
```
Comment (mdiep) [inline Tests/TentacleTests/FileTests.swift]: 🤔

**Review by mdiep — CHANGES_REQUESTED** [review]
Really neat to see this with the built-in Swift `Decodable` stuff! 👏

File: .gitmodules [hunk 15]
```diff
@@ -1,15 +1,8 @@
...
-[submodule "Carthage/Checkouts/Curry"]
-	path = Carthage/Checkouts/Curry
-	url = https://github.com/thoughtbot/Curry.git
 [submodule "Carthage/Checkouts/ReactiveSwift"]
 	path = Carthage/Checkouts/ReactiveSwift
 	url = https://github.com/ReactiveCocoa/ReactiveSwift.git
-[submodule "Carthage/Checkouts/Runes"]
-	path = Carthage/Checkouts/Runes
-	url = https://github.com/thoughtbot/Runes.git
+[submodule "Carthage/Checkouts/result"]
```
Comment (dcaunt) [inline .gitmodules]: Result is now duplicated in this file, and (nit) it should be capitalized `Result`

**Comment by Palleas** [comment]
Thank you all for the feedback, I'll take care of most of them this week. The rollback of SE-0110 should be in the next Xcode seed right?

**Comment by mdiep** [comment]
> The rollback of SE-0110 should be in the next Xcode seed right?

I think so, yes. Otherwise you can probably download a snapshot from swift.org.

**Comment by Palleas** [comment]
Xcode 9b3 reverts SE-110 so I'll tackle all the feedback this week!

File: Sources/Tentacle/Client.swift [hunk 16]
```diff
@@ -6,11 +6,12 @@
 //  Copyright © 2016 Matt Diephouse. All rights reserved.
 //
 
-import Argo
 import Foundation
 import ReactiveSwift
 import Result
 
+//extension Array where Element == ResourceType : ResourceType {}
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: 😞

File: Sources/Tentacle/Client.swift [hunk 17]
```diff
@@ -26,8 +27,8 @@ extension URL {
     
     internal init<Value>(_ server: Server, _ request: Request<Value>, page: UInt? = nil, perPage: UInt? = nil) {
         let queryItems = [ ("page", page), ("per_page", perPage) ]
-            .filter { _, value in value != nil }
-            .map { name, value in URLQueryItem(name: name, value: "\(value!)") }
+            .filter { (_, value) -> Bool in value != nil }
+            .map { (name, value) -> URLQueryItem in URLQueryItem(name: name, value: "\(value!)") }
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: You should be able to remove these return types?

File: Sources/Tentacle/Client.swift [hunk 18]
```diff
@@ -245,20 +236,23 @@ public final class Client {
...
+                    .mapError { Error.jsonDecodingError($0) }
             }
+            .flatMap(.concat, { (response, data) -> SignalProducer<(Response, [Resource]), Error> in
+                let current = SignalProducer<(Response, [Resource]), Error>(value: (response, data))
+                guard let _ = response.links["next"] else {
+                    return current.concat(.empty)
+                }
+
+                return current.concat(self.execute(request, page: nextPage, perPage: perPage))
+            })
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: Trailing closure syntax here, please. 🙏

File: Sources/Tentacle/Client.swift [hunk 19]
```diff
@@ -272,7 +266,7 @@ extension Client.Error: Hashable {
             return (error1 as NSError) == (error2 as NSError)
 
         case let (.jsonDecodingError(error1), .jsonDecodingError(error2)):
-            return error1 == error2
+            return error1 == error2 // FIXME
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: Why is this labeled `FIXME`?

File: Sources/Tentacle/Tree.swift [hunk 20]
```diff
@@ -81,6 +143,47 @@ public struct Tree: CustomStringConvertible {
...
+        }
+
+        private enum CodingKeys: String, CodingKey {
+            case type
+            case sha
+            case path
+            case mode
+            case url
+            case size
+        }
```
Comment (mdiep) [inline Sources/Tentacle/Tree.swift:157]: This shouldn't be needed since all the cases have the default value?

**Review by mdiep — CHANGES_REQUESTED** [review]

**Comment by mdiep** [comment]
I'm liking these changes!

If it'd be helpful, we could add `ID`/`Identifiable` in a separate PR before this one.

File: Sources/Tentacle/Client.swift (hunk 17 above)
Comment (Palleas) [reply Sources/Tentacle/Client.swift]: Done!

File: Sources/Tentacle/Client.swift (hunk 18 above)
Comment (Palleas) [reply Sources/Tentacle/Client.swift]: Not sure how I missed it, fixed!

File: Sources/Tentacle/Tree.swift (hunk 20 above)
Comment (Palleas) [reply Sources/Tentacle/Tree.swift:157]: If think they are because url and size are not exposed as properties and we are using them for encoding, but maybe I'm missing something..?

**Comment by dcaunt** [comment]
How's this going Palleas, anything I can help with? I was also wondering if it would be worthwhile adding the Xcode 9 job in a separate PR.

**Comment by dcaunt** [comment]
How's this going @Palleas, anything I can help with? I was also wondering if it would be worthwhile adding the Xcode 9 job in a separate PR.

**Comment by Palleas** [comment]
@dcaunt Almost done, I'll try to find some time soon. An Xcode 9 job would help a lot for sure!

**Comment by dcaunt** [comment]
Added in #88

**Comment by Palleas** [comment]
@dcaunt @mdiep I think I'm done..? 😱

**Comment by dcaunt** [comment]
@Palleas awesome! I'll do my best to review it this week.

File: Sources/Tentacle/Label.swift [hunk 21]
```diff
@@ -7,17 +7,31 @@
...
+        self.color = Color(hex: try container.decode(String.self, forKey: .color))
+    }
+
+    public init(name: String, color: Color) {
+        self.name = name
+        self.color = color
+    }
+
+    // TODO: remove this, or replace `Color` with something (De)codable
+    private enum CodingKeys: String, CodingKey {
```
Comment (dcaunt) [inline Sources/Tentacle/Label.swift]: This relates to my comment on `Author`. Should we remove definitions of `CodingKeys` that would match those that would be generated by the compiler, or is it safer to be explicit?

@mdiep had originally suggested we remove these.

File: Sources/Tentacle/Author.swift [hunk 22]
```diff
@@ -19,14 +18,18 @@ public struct Author {
...
+    }
 }
 
 extension Author: Encodable {
-    public func encode() -> JSON {
-        return JSON.object([
-            "name": .string(name),
-            "email": .string(email)
-        ])
+    public func encode(to encoder: Encoder) throws {
```
Comment (dcaunt) [inline Sources/Tentacle/Author.swift]: If we declare `Encodable` conformance on `Author` rather than in an extension, the compiler can synthesize the `encode` method and we can use the synthesized `CodingKeys`.

**Comment by dcaunt** [comment]
@Palleas looks like tests are [failing to compile](https://travis-ci.org/mdiep/Tentacle/jobs/275378798#L838) under Release configuration

**Comment by Palleas** [comment]
Weird, I’ll have a look tomorrow!

File: Sources/Tentacle/Author.swift (hunk 22 above)
Comment (Palleas) [reply Sources/Tentacle/Author.swift]: Good point!

File: Sources/Tentacle/Label.swift (hunk 21 above)
Comment (Palleas) [reply Sources/Tentacle/Label.swift]: I can't reference CodingKeys in the `init` unless I make a custom `Color` type that's Decodable (instead of just being an alias to `UI/NSColor`. I'll add that in a minute.

Comment (dcaunt) [reply Sources/Tentacle/Label.swift]: Thinking about this further, I think we should be explicit with `CodingKeys` to avoid any breakage in future refactoring. If Matt has different ideas we can always remove them in a future PR.

**Comment by dcaunt** [comment]
@Palleas The tests are failing on iOS due to the `SWIFT_VERSION` build setting being set to `4.0`. For the OS X framework it's set to `3.2`.

**Comment by Palleas** [comment]
Weird that the migration tool didn’t update it. Any chance you can make the change? I’m on vacations for the next 2 weeks and I don’t have my laptop.

**Comment by dcaunt** [comment]
Sure, I can pick it up. Have a great vacation!

**Comment by Palleas** [comment]
Thank! If there is too much to do feel free to send it back to me!

**Comment by ikesyo** [comment]
@Palleas @dcaunt CI is green now with some updates.

ae7e37f...dbb9a21

File: Package.swift [hunk 23]
```diff
@@ -1,9 +1,17 @@
+// swift-tools-version:4.0
```
Comment (dcaunt) [inline Package.swift:1]: Ah, this is what I was missing.

**Comment by dcaunt** [comment]
@ikesyo Thanks for fixing the remaining issues!

Do you think we should explicitly define the `CodingKeys` as discussed above? I know you were keen to take this approach in some other projects.

**Comment by ikesyo** [comment]
> Do you think we should explicitly define the `CodingKeys` as discussed above?

I like the explicitness atm but either is okay for me.

File: Sources/Tentacle/Homepage.swift [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+//  Homepage.swift
+//  Tentacle
+//
+//  Created by Romain Pouclet on 2017-06-20.
+//  Copyright © 2017 Matt Diephouse. All rights reserved.
+//
+
+import Foundation
+
+public enum Homepage: Decodable {
```
Comment (ikesyo) [inline Sources/Tentacle/Homepage.swift]: What is the purpose of this type? Keep using `URL` seems better to me.

Comment (dcaunt) [reply Sources/Tentacle/Homepage.swift]: Unless the API is expected to respond with an arbitrary String and not a valid URL, I agree.

Comment (ikesyo) [reply Sources/Tentacle/Homepage.swift]: I understand now that the problem here is empty string handling (`""`). Empty string is not valid input even if a property is `URL?`.

Comment (dcaunt) [reply Sources/Tentacle/Homepage.swift]: In that case it seems like we should revert to `URL?` and change the decoding?

Comment (mdiep) [reply Sources/Tentacle/Homepage.swift]: GitHub will let you set this field to any string.

BUT your github.com profile will only show the string if it's actually a URL. So I think we could revert to a `URL` and just set it to `nil` if it's not a valid URL.

Comment (dcaunt) [reply Sources/Tentacle/Homepage.swift]: Pushed aa43141 to address this.

**Comment by Palleas** [comment]
About the Homepage change, bear in mind that it’s not necessarily a valid URL, there is no validation on Github’s side. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: Sources/Tentacle/Homepage.swift (hunk 24 above)
Comment (Palleas) [reply Sources/Tentacle/Homepage.swift]: Oops, missed this, removed my comment 🙊
Are we sure it’s the same with a repository? I remember having this conversation a while ago 🤔

File: Sources/Tentacle/Tree.swift [hunk 25]
```diff
@@ -270,9 +270,6 @@ internal struct NewTree: Encodable {
     func encode(to encoder: Encoder) throws {
         var container = encoder.container(keyedBy: CodingKeys.self)
         try container.encode(entries, forKey: .entries)
-
-        if let base = base {
-            try container.encode(base, forKey: .base)
-        }
```
Comment (Palleas) [inline Sources/Tentacle/Tree.swift]: TIL!

File: Sources/Tentacle/Client.swift [hunk 26]
```diff
@@ -245,19 +228,22 @@ public final class Client {
...
+                    .map { (response, $0) }
                     .mapError(Error.jsonDecodingError)
             }
-            .flatMap(.concat) { response, json -> SignalProducer<(Response, [Resource]), Error> in
-                return SignalProducer(value: (response, json))
-                    .concat(response.links["next"] == nil ? SignalProducer.empty : self.execute(request, page: nextPage, perPage: perPage))
+            .flatMap(.concat) { response, data -> SignalProducer<(Response, [Resource]), Error> in
+                let current = SignalProducer<(Response, [Resource]), Error>(value: (response, data))
+                guard let _ = response.links["next"] else {
+                    return current.concat(.empty)
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: `.concat(.empty)` is effectively a noop. This can just be `return current`.

File: Sources/Tentacle/Commit.swift [hunk 27]
```diff
@@ -7,13 +7,19 @@
...
+
+        public init(from decoder: Decoder) throws {
+            let container = try decoder.singleValueContainer()
+            self.raw = try container.decode(String.self)
+        }
+    }
+
     /// SHA of the commit
-    public let sha: SHA
+    public let sha: Hash
```
Comment (mdiep) [inline Sources/Tentacle/Commit.swift]: Why did this change from a `SHA`?

File: Sources/Tentacle/Commit.swift [hunk 28]
```diff
@@ -30,15 +36,15 @@ public struct Commit {
...
     public let parents: [Parent]
 
-    public struct Parent {
+    public struct Parent: ResourceType {
         /// URL to see the parent commit in a browser
         public let url: URL
 
         /// SHA of the parent commit
-        public let sha: SHA
+        public let sha: Hash
```
Comment (mdiep) [inline Sources/Tentacle/Commit.swift]: Why did this change from a `SHA`?

File: Sources/Tentacle/Content.swift [hunk 29]
```diff
@@ -57,6 +54,43 @@ public enum Content {
...
+                    let url = try container.decodeIfPresent(URL.self, forKey: .downloadURL)
+                    self = .symlink(target: target, downloadURL: url)
+                default:
+                    throw DecodingError.dataCorruptedError(
+                        forKey: CodingKeys.type,
+                        in: container,
+                        debugDescription: "Invalid content-type \(type)"
+                    )
+                }
+            }
```
Comment (mdiep) [inline Sources/Tentacle/Content.swift]: 🙈

File: Sources/Tentacle/JSONExtensions.swift [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+        let decoder = JSONDecoder()
+        decoder.dateDecodingStrategy = .formatted(DateFormatter.iso8601)
+        return try decoder.decode([T].self, from: payload)
+    }
+}
+
+extension DecodingError: Equatable {
+    static public func ==(lhs: DecodingError, rhs: DecodingError) -> Bool {
+        switch (lhs, rhs) {
+        default: return false // FIXME
```
Comment (mdiep) [inline Sources/Tentacle/JSONExtensions.swift]: FIXME

File: Sources/Tentacle/Label.swift (hunk 21 above)
Comment (mdiep) [reply Sources/Tentacle/Label.swift]: I'm fine with being explicit. 👍

File: Sources/Tentacle/Label.swift [hunk 31]
```diff
@@ -7,17 +7,31 @@
...
+        self.name = try container.decode(String.self, forKey: .name)
+        self.color = Color(hex: try container.decode(String.self, forKey: .color))
+    }
+
+    public init(name: String, color: Color) {
+        self.name = name
+        self.color = color
+    }
+
+    // TODO: remove this, or replace `Color` with something (De)codable
```
Comment (mdiep) [inline Sources/Tentacle/Label.swift]: I think this comment can go away?

File: Sources/Tentacle/RepositoryInfo.swift [hunk 32]
```diff
@@ -63,6 +60,65 @@ public struct RepositoryInfo: CustomStringConvertible, Identifiable {
...
     }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        self.id = try container.decode(ID.self, forKey: .id)
+        self.owner = try container.decode(UserInfo.self, forKey: .owner)
+        self.name = try container.decode(String.self, forKey: .name)
+        self.nameWithOwner = try container.decode(String.self, forKey: .nameWithOwner)
+        self.body = try? container.decode(String.self, forKey: .body)
```
Comment (mdiep) [inline Sources/Tentacle/RepositoryInfo.swift]: I think this should be `try container.decodeIfPresent`?

**Review by mdiep — CHANGES_REQUESTED** [review]
Looks good overall! Thanks for tackling this! It's no small effort. 🎉

File: Sources/Tentacle/Commit.swift (hunk 27 above)
Comment (Palleas) [reply Sources/Tentacle/Commit.swift]: Because it’s not exactly the same type IIRC

File: Sources/Tentacle/RepositoryInfo.swift (hunk 32 above)
Comment (dcaunt) [reply Sources/Tentacle/RepositoryInfo.swift]: Yeah, it should 😞

File: Sources/Tentacle/Tree.swift [hunk 33]
```diff
@@ -54,20 +53,84 @@ public struct Tree: CustomStringConvertible {
...
 
-    public struct Entry {
+    private enum CodingKeys: String, CodingKey {
+        case sha
+        case url
+        case entries = "tree"
+        case isTruncated = "truncated"
+    }
+
+    public struct SHA: Codable {
```
Comment (mdiep) [inline Sources/Tentacle/Tree.swift]: We shouldn't need this one either. A single `SHA` type should be enough.

File: Sources/Tentacle/RepositoryInfo.swift (hunk 32 above)
Comment (Palleas) [reply Sources/Tentacle/RepositoryInfo.swift]: Fixed.

File: Sources/Tentacle/JSONExtensions.swift (hunk 30 above)
Comment (Palleas) [reply Sources/Tentacle/JSONExtensions.swift]: Any idea how to do proper Equatable implementation for `DecodingError`? 🤔

File: Sources/Tentacle/Content.swift (hunk 29 above)
Comment (Palleas) [reply Sources/Tentacle/Content.swift]: How else would you handle this? 🤔

**Comment by Palleas** [comment]
I remember why there are 2 SHAs:

On `Branch`, we have:
* Branch
    * Commit
        * Sha (String)

But on `Tree`, we have: 
* Tree
    * Sha (String)

I'll change the way it works for `Branch`, it feels like a hack.

File: Sources/Tentacle/JSONExtensions.swift (hunk 30 above)
Comment (Palleas) [reply Sources/Tentacle/JSONExtensions.swift]: I have a proposition suggested to me in Slack in f9d214e, let me know what you think!

File: Sources/Tentacle/Content.swift (hunk 29 above)
Comment (Palleas) [reply Sources/Tentacle/Content.swift]: Made a proposal in 261c768, let me know what you think. At least we're no longer switching over strings.

File: Sources/Tentacle/Client.swift (hunk 26 above)
Comment (Palleas) [reply Sources/Tentacle/Client.swift]: Fixed!

File: Sources/Tentacle/RepositoryInfo.swift [hunk 34]
```diff
@@ -68,9 +68,9 @@ public struct RepositoryInfo: CustomStringConvertible, ResourceType, Identifiabl
         self.owner = try container.decode(UserInfo.self, forKey: .owner)
         self.name = try container.decode(String.self, forKey: .name)
         self.nameWithOwner = try container.decode(String.self, forKey: .nameWithOwner)
-        self.body = try? container.decode(String.self, forKey: .body)
+        self.body = try container.decodeIfPresent(String.self, forKey: .body)
         self.url = try container.decode(URL.self, forKey: .url)
-        self.homepage = try? container.decode(URL.self, forKey: .homepage)
+        self.homepage = try container.decodeIfPresent(URL.self, forKey: .homepage)
```
Comment (ikesyo) [inline Sources/Tentacle/RepositoryInfo.swift]: This doesn't work for empty string. That should be why `try?` was used here.

File: Sources/Tentacle/RepositoryInfo.swift [hunk 35]
```diff
@@ -63,6 +60,65 @@ public struct RepositoryInfo: CustomStringConvertible, Identifiable {
...
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        self.id = try container.decode(ID.self, forKey: .id)
+        self.owner = try container.decode(UserInfo.self, forKey: .owner)
+        self.name = try container.decode(String.self, forKey: .name)
+        self.nameWithOwner = try container.decode(String.self, forKey: .nameWithOwner)
+        self.body = try container.decodeIfPresent(String.self, forKey: .body)
+        self.url = try container.decode(URL.self, forKey: .url)
+        self.homepage = try container.decodeIfPresent(URL.self, forKey: .homepage)
```
Comment (ikesyo) [inline Sources/Tentacle/RepositoryInfo.swift]: This doesn't work for empty string. That should be why `try?` was used here.

Comment (mdiep) [reply Sources/Tentacle/RepositoryInfo.swift]: That doesn't totally match the intent, since a string that's not a URL should be `nil`, but a number should throw. But it's probably close enough. 🤷‍♂️

Comment (Palleas) [reply Sources/Tentacle/RepositoryInfo.swift]: Soooo, back to `try?` is what you meant?

Comment (mdiep) [reply Sources/Tentacle/RepositoryInfo.swift]: Yeah, just for `homepage`.

Comment (Palleas) [reply Sources/Tentacle/RepositoryInfo.swift]: Done!

**Comment by Palleas** [comment]
Also fixed the remaining failing test (in Tree encoding)

File: Sources/Tentacle/Branch.swift [hunk 36]
```diff
@@ -20,36 +17,42 @@ extension Repository {
...
+        public let sha: SHA
+    }
 
     /// Name of the branch
     public let name: String
 
-    /// Sha of the commit the branch points to
-    public let sha: SHA
+    /// The commit the branch points to
+    public let commit: Commit
```
Comment (mdiep) [inline Sources/Tentacle/Branch.swift:30]: I'm still confused about why this isn't just a `SHA`. 🙈

File: Sources/Tentacle/Client.swift [hunk 37]
```diff
@@ -188,55 +181,45 @@ public final class Client {
...
                     }
-                    .map { json in
-                        return (Response(headerFields: headers), json)
+
+                    if response.statusCode >= 400 && response.statusCode < 600 {
+                        return decode(data)
+                            .mapError(Error.jsonDecodingError)
+                            .flatMap { error in
+                                .failure(Error.apiError(response.statusCode, Response(headerFields: headers), error))
+                        }
```
Comment (mdiep) [inline Sources/Tentacle/Client.swift]: This `}` is under indented.

File: Sources/Tentacle/Commit.swift (hunk 27 above)
Comment (mdiep) [reply Sources/Tentacle/Commit.swift]: Can you elaborate on this? I think it should be the same thing.

**Review by mdiep — COMMENTED** [review]
The question of the `SHA` types is really the only remaining thing.

File: Sources/Tentacle/Branch.swift (hunk 36 above)
Comment (Palleas) [reply Sources/Tentacle/Branch.swift:30]: This is what the payload returns: 

```json
[
  {
    "name": "master",
    "commit": {
      "sha": "6dcb09b5b57875f334f61aebed695e2e4193db5e",
      "url": "https://api.github.com/repos/octocat/Hello-World/commits/c5b97d5ae6c19d5c5df71a34c7fbeeda2479ccbc"
    },
    "protected": true,
    "protection_url": "https://api.github.com/repos/octocat/Hello-World/branches/master/protection"
  }
]
```

I can retrieve the `sha` directly with a custom `init`, is that what you have in mind? Maybe I'm just missing something  🤷‍♂️

Comment (mdiep) [reply Sources/Tentacle/Branch.swift:30]: Ah, okay! I didn't realize there was another property returned on `commit`! 😅

**Review by mdiep — APPROVED** [review]

**Comment by mdiep** [comment]
This is great! Thank you SO SO SO much for doing this. ✨ It's great to see what Swift's built-in decoding looks like, warts and all.

Sorry about the long review times and my general cluelessness in this PR. Things have been a little crazy here. 😅
